lib/bolt/last_failure.rb

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require 'json'
+
+module Bolt
+  class LastFailure
+    def initialize(path)
+      @path = path
+    end
+
+    def data
+      @data ||= JSON.parse(File.read(@path))
+      unless @data.is_a?(Array) && @data.all? { |r| r['target'] && r['status'] }
+        raise Bolt::FileError.new("Missing data in last_failure file: #{@path}", @path)
+      end
+      @data
+    rescue JSON::ParserError
+      raise Bolt::FileError.new("Could not parse last_failure file: #{@path}", @path)
+    rescue IOError, SystemCallError
+      raise Bolt::FileError.new("Could not read last_failure file: #{@path}", @path)
+    end
+
+    def get_targets(filter)
+      filtered = case filter
+                 when 'all'
+                   data
+                 when 'failure'
+                   data.select { |result| result['status'] == 'failure' }
+                 when 'success'
+                   data.select { |result| result['status'] == 'success' }
+                 else
+                   raise Bolt::CLIError, "Unexpected option #{filter} for '--retry'"
+                 end
+      filtered.map { |result| result['target'] }
+    end
+
+    def write(result_set)
+      data = result_set.map { |result| result.status_hash.select { |k, _| %i[target status].include? k } }
```

Comment:
This will result in the password for nodes where the password is in the name being persisted to disk. The only way this seems to result in new information being persisted to disk is if:
1. The password was included in `--nodes` which was probably already persisted by the shell history unless the nodes were passed on `stdin`
2. Node uris are built inside a plan and contain the password.

Options:
1. We can have a config option to disable writing this file. This may be a good idea anyway
2. We can formalize the concept of a "safe" name in the target and save that with the knowledge that those nodes can't be rerun.
3. do nothing.

Reply:
I've update the PR with 2 so the remaining question is do we need to do 1 in addition.